### ISSUE TITLE
Performance of concurrent Request/Reply

### ISSUE THREAD

**Issue by plaisted** [nats-io/nats.net.v1#202]
I've was working on a basic RPC setup with NATS.Client and found large performance degradation using the built in Request feature if the connection is shared for multiple concurrent requests.

The basic code per concurrent request stream is essentially:
```csharp
        public static async Task Run(int count, IConnection connection, ManualResetEvent starter)
        {
            if (starter != null)
            {
                // ensure connection warm and wait
                await Run(10, connection, null);
                starter.WaitOne();
            }
            var sw = new Stopwatch();
            sw.Start();
            for (var i = 0; i < count; i++)
            {
                var result = await connection.RequestAsync("test", Encoding.UTF8.GetBytes("ping"));
            }
            sw.Stop();
            Console.WriteLine("Per request: " + Decimal.Divide(sw.ElapsedMilliseconds, count));
        }
```

If only one task is spawned I get an average round trip of under 1ms (0.6-0.7 typically).

If I spawn 10 tasks running, the average is approximately 10x higher (~7ms).

This makes it seem like the requests are unable to run concurrently at all. I would expect some performance degradation (they are sharing the same TCP connection after all) but a linear increase like this seems off. This suggests a single RPC is somehow saturating the connection (doesn't seem possible), or my code (or the library) is blocking in some inefficient way.

I've tested the same type of setup sharing the same network connection with no issues using GRPC or RabbitMQ. The performance dropoff was gradual and could get quite a large number concurrently.

Any suggestions on how to handle a concurrent requests like this? The final application is a web server publishing messages so the concurrency is driven by requests coming in.

**Comment by ColinSullivan1** [comment]
@plaisted , thank you for raising this!  Your code looks straightforward to me; I'll take a look.  Which version are you running?

**Comment by plaisted** [comment]
I installed the latest Nuget earlier today. I'll check later when I have access to the servers I was testing on. Thanks for quick response.

**Comment by plaisted** [comment]
Running on Windows 10x64.
nats-server version 1.0.4
NATS.Client 0.8.0
.Net: netcoreapp2.0

**Comment by ColinSullivan1** [comment]
Absolutely.  Thanks for the info!  I'll be out of my office over the weekend so this may take a few days, but I'll try to get to this ASAP.

**Comment by plaisted** [comment]
Very simple async subscriber with a publish/flush mirroring an example. I
tried multiple subscriber apps on the same queue at the same time without
significant change in round trip time so appears to be on the request side.

I'll try to find time this weekend to throw together a full example.

**Comment by ColinSullivan1** [comment]
@plaisted, I have a few suggestions for your test.

First, Could you aggregate the tasks and then await on them outside of the loop?  Here it looks like you'll be waiting on each request before making the next.  Some code that does that can be found [here](https://github.com/nats-io/csharp-nats/blob/v0.8.0/NATSUnitTests/UnitTestBasic.cs#L510).  

Also, do you have enough subscribers servicing the requests?  If you setup a number of queue subscribers to process requests in parallel , you'll have more requests being processed simultaneously.  If you have only one subscriber, it'll be a bottleneck increasing latency, effectively making the test run serially.  Does this make sense?

**Comment by plaisted** [comment]
FYI I found a workaround but will response to your specific questions.

> First, Could you aggregate the tasks and then await on them outside of the loop? Here it looks like you'll be waiting on each request before making the next. Some code that does that can be found here.

The waiting on each request is on purpose. I loop through a few thousand requests to get an average on the round trip time. The concurrency comes in by spawning multiple of the shown `Run()` Tasks at the same time. I've been using this same style of benchmarking for the other frameworks I've tested, it's to mimic the artillery.io HTTP load testing methods (which I use for HTTP endpoint loading).

> Also, do you have enough subscribers servicing the requests? If you setup a number of queue subscribers to process requests in parallel , you'll have more requests being processed simultaneously. If you have only one subscriber, it'll be a bottleneck increasing latency, effectively making the test run serially. Does this make sense?

Yes it does. I've experimented with adding additional workers to make sure there aren't any bottlenecks there and it had little affect.

Any now the workaround I found:

Running each request (as I was doing) like:
```csharp
var result = await connection.RequestAsync("test", Encoding.UTF8.GetBytes("ping"));
```
somehow causes a deadlock for me when running multiple concurrent request Tasks. Certain instances of the `Run()` method just hang until others have finished (don't even complete the "warmup" phase on the task).

Simple adding a timeout to the request like,
```csharp
var result = await connection.RequestAsync("test", Encoding.UTF8.GetBytes("ping"), 100);
```
solves the issues and everything performs as I'd expect and all `Run()` methods run concurrently. Note this timeout is never triggered (no exceptions). I haven't had time to dig into the cause of this.

I've dropped the code used at https://github.com/plaisted/rpc-perf-test

General results all running (1 x requester, 2 x responder, nats server) on my surface book locally:
1 worker threads: 3051 RPS, 0.32ms round trip
2 worker threads: 5005 RPS, 0.4ms round trip
3 worker threads: 5133 RPS, 0.58ms round trip (get close to 100% CPU here)
4 worker threads: 5334 RPS, 0.56ms round trip

**Comment by ColinSullivan1** [comment]
@plaisted ,  Ahh, I see - thanks for explaining the code snippet.  I'm glad you found that workaround.  

This is really good info; I appreciate you providing all of this and the sample code.  It's extremely helpful!  I'll dig in as soon as I can.

Thanks,
Colin

**Comment by ashalabad** [comment]
I just ran provided tests with new client version 0.8.1 and results are quite alarming for 10 workers:
Single ave: 14.347
Success: 99.700%
Single ave: 14.666
Success: 99.800%
Single ave: 14.713
Success: 99.700%
Single ave: 14.783
Success: 99.700%
Single ave: 14.79
Success: 99.600%
Single ave: 14.788
Success: 99.700%
Single ave: 14.866
Success: 99.700%
Single ave: 14.881
Success: 99.700%
Single ave: 14.885
Success: 99.700%
Single ave: 14.88
Success: 99.600%
Per request : 14.7599 ms.
Total RPS: 671.63677883000873127812479010.

I see serious performance degradation.

Thanks
Igor

**Comment by ashalabad** [comment]
It's NET Core 2.1 , i7-6500U, 16Gb 2.5GHZ

**Comment by ColinSullivan1** [comment]
This should be resolved by #339, and is now in a pre-release, https://www.nuget.org/packages/NATS.Client/0.10.1-pre1
